ci/scripts/check-bazel-tags.sh

```diff
@@ -6,14 +6,125 @@
-if [[ ${untagged} ]]; then # Check that all targets that depend on vivado are tagged
-  echo "Target(s): ${untagged} depend(s) on a bitstream_splice but isn't tagged with vivado or manual";
+check_empty () {
+if [[ ${1} ]]; then
+  echo "Error:"
+  echo "$1"|sed 's/^/    /';
+  echo "$2"
   exit 1
 fi
+}
+
+# This check ensures OpenTitan software can be built with a wildcard without
+# waiting for Verilator using --build_tag_filters=-verilator
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      //hw:verilator
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(verilator|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on //hw:verilator, please tag it with verilator or
+(to prevent matching any wildcards) manual.
+NOTE: test_suites that contain targets with different tags should almost
+universally use the manual tag."
+
+# This check ensures Opentitan software can be built with wildcards in
+# environments that don't have vivado or vivado tools installed by using
+# --build_tag_filters=-vivado.
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      kind(
+          'bitstream_splice',
+          //...
+      )
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(cw310_rom|cw310_test_rom|vivado|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on a bitstream_splice that isn't cached.
+Please tag it with vivado or (to prevent matching any wildcards) manual.
+NOTE: test_suites that contain targets with different tags should almost
+universally use the manual tag."
+
+# This check ensures cw310 users may group tests that depend on the cached
+# cw310_test_rom bitstream.
+mistagged=$(./bazelisk.sh query \
+    "tests(`# Only output tests`
+        rdeps(
+            attr(`# Anything tagged cw310_test_rom`
+                tags,
+                '$(exact_regex cw310_test_rom)',
+                tests(//...)
+            ),
+            kind(`# That depends on a bitstream splice`
+                'bitstream_splice',
+                //...
+            )
+            except`# Other than those used to build the test_ROM bitstream`
+            deps(
+                //hw/bitstream:test_rom
+            )
+        )
+    )" \
+    --output=label_kind)
+check_empty "${mistagged}" \
+"Target(s) above depend(s) on a bitstream_splice rule other than those used to
+generate the cached bitstream with the test_ROM, but is tagged with cw310_test_rom.
+Please either:
+-Correct the dependencies to exclude other bitstream_splices,
+-Correct the target by setting it to something other than cw310_test_rom,
+-Remove the cw310_test_rom tag.
+NOTE: test_suites that contain targets with different tags should almost
+universally use the manual tag."
+
+
+
+# This check ensures cw310 users may group tests that depend on the cached
+# cw310_rom bitstream.
```

Comment:
I'm having trouble interpreting this comment relative to the query below.

It looks to me that we're searching for mistagged tests. The check fails if we find a test that depends on a `bitstream_splice` and is tagged `cw310_rom`, provided that the `bitstream_splice` is not used by `//hw/bitstream:rom`.

Reply:
This comment describes why we do the check. The query itself is documented below. I can describe it without using the tags to name the bitstreams.

Reply:
@dmcardle I don't know what you're asking for here so I'm blocked on resolving it.

Reply:
Maybe I'm just tired, but I still find the comment hard to parse. In particular, "with the cw310_rom tag" feels like it's too far away from the verb "group". Also, I wasn't sure what "the cached ROM bitstream" referred to at first. Anyway, same as above, I don't want to block the PR on this.

Reply:
> Maybe I'm just tired, but I still find the comment hard to parse. In particular, "with the cw310_rom tag" feels like it's too far away from the verb "group".

Notionally or as written in the sentence?

Reply:
> "the cached ROM bitstream"

What should we call it? I've been calling it cw310_rom which has also caused problems and *.spliced might not even be in use anymore...
Names matter and using the wrong one here will cost more in the long run than fixing it (plus I still don't have approval yet so it's not like I have an excuse to merge it in and force the issue)

"the cw310 bitstream with the ROM spliced into it and hosted in GCP" feels wordy, but if that's what it takes to be clear, I might as well get it right in a comment. We could hyphenate it so people couldn't call it wordy, and because hyphenating long strings is a good schtick.

Reply:
```
# This check ensures cw310 users may filter tests with the "cw310_rom" tag to
# run tests with a common bitstream available in GCP to avoid issues building
# and loading new bitstreams
```
?

Reply:
Works for me :)

If we haven't already, we should probably document this optimization at some point. Then we can just point to the doc whenever this comes up, rather than trying to re-explain why we're grouping bitstream. We'd want to update `azure-pipelines.yml`, `ci/scripts/run-fpga-cw310-tests.sh`, and this file — maybe I'm missing some.